### DISCUSSION TITLE
Let setting UI support for other program languages

### DISCUSSION THREAD

**Discussion by Zeroto521** [Flow-Launcher/Flow.Launcher#514] [Ideas]
## Question

As title.

More can see [jjw24/Wox#178](https://github.com/jjw24/Wox/issues/178).

## Discussion

I think just like using **JSON-RPC** for supporting other program languages. 

Why not using some **template file** for support setting UI. 

But at the same time, we need a **tool** to compile the file and  **inject in** the Flow main program.

To make it just like a `C#` plugin.

![Interaction](https://user-images.githubusercontent.com/25895405/93783646-85321a00-fc5e-11ea-8f8b-19f55721d03b.png)

**Comment by jjw24** [comment]
Interesting indeed. Let me think how we can do this

**Reply by ghost** [reply]
Idk if this is good but i found from this [Quora question](https://www.quora.com/How-can-I-use-C-to-create-a-UI-for-a-Python-program-What-are-all-possibilities-and-methods)

Maybe you can get [IronPython](https://ironpython.net/) to be able to make GUI with xaml and code using python

**Reply by ghost** [reply]
You can write UI in xaml and interact code with python using IronPython

**Reply by taooceros** [reply]
That may require a quite large refactor since it needs assembly interaction, but currently python plugin is only communicated via args (with jsonrpc protocol).

**Reply by taooceros** [reply]
But IronPython seems quite interesting. Probably python developer can use it to assembly in python and let Flow load the assembly as dotnet plugin, which is almost the same as C# plugin.

**Reply by ghost** [reply]
Yeah the ui can be written in xaml just like c# but the code for the plugin is written in python which will enable python coders to make python plugin with UI just like c# 😁

**Reply by ghost** [reply]
Maybe take a look on this [YouTube tutorial](https://youtu.be/i4lSlvALngw)

**Reply by Zeroto521** [reply]
In my memory, IronPython is an old-time interpreter.
Now it only supports py3.4 and just in an alpha version.
So I wouldn't consider it at present.

![image](https://user-images.githubusercontent.com/25895405/125731879-8570afe6-4856-47c4-99d6-4fa859774d0c.png)

**Reply by pc223** [reply]
I'm not too sure why we need IronPython? 💭 You want plugins have their own UI? Interesting but a bit overkill for now I think 😅
 
About the settings problem, I think let plugins able to add more options in settings-page is a really cool idea, not sure how to implement tho. I think we can have couple of basic cases: string-config, boolean-config, list-choices-config (combobox)... And we could try to map these with the settings UI and plugins

One case study we can look into is VSCode, I see they have some kind of map from `settings.json` file to the settings UI

<img src="https://user-images.githubusercontent.com/10551242/126071391-a1eeac18-9994-4802-8e9f-b2d0f798b23a.png" width=600>

**Reply by ghost** [reply]
I mean it would be nice if python plugins can have settings UI like a C# plugin can
[Image](https://i.imgur.com/JI3QxxE.png) example

**Reply by taooceros** [reply]
> I mean it would be nice if python plugins can have settings UI like a C# plugin can
> [Image](https://i.imgur.com/JI3QxxE.png) example

Yeah I think so. It would make python plugin be able to do more things.

I think @Zeroto521 's proposal is possible, but we need a complete proposal to implement it.

**Comment by taooceros** [comment]
## Proposal

### Basic Property

Plugins need to provide a poco json to indicate what kind of config they want to have.
```json
{
    "property1": "default value",
    "property2": "default value"
}
```
~~JSONRPCClass need to parse the value and create a poco class (or dynamic parse)~~ Parse it to a Dictionary, and stored it into a json file.
Plugin will access setting via the json file.

### UI Element

Key will create a `Label` with key name (`TextBlock` can also be used here)
String, Int, Double will be treated as string and create a `textbox`
Bool will create a `CheckBox`

### Unfinished

- [ ] Support for saving List of value
    - [ ] UI control
```json
{
    "list1": ["default value 1", "default value 2"]
}
```
- [ ] Support for saving Dictionary or class
    - [ ] UI control
```json
{
    "dic1": {"key": "value" }
}
```

TODO:
- [ ] Dynamic create usercontrol based on poco.
- [ ] Parse the json input to a poco

**Reply by jjw24** [reply]
Haven't thought this through thoroughly but would it be better to use yaml instead of json?

**Reply by taooceros** [reply]
> Haven't thought this through thoroughly but would it be better to use yaml instead of json?

This needs to be json because it is a request response. The config file can be yaml, but why yaml over json?

**Reply by Zeroto521** [reply]
Maybe we design from UI first?

To know about how many **components** we could use, then design to correspond data structure and file format.

**Reply by taooceros** [reply]
I have written a draft using `Label, TextBlock or CheckBox`

**Reply by taooceros** [reply]
Simplified a bit. No need to manually type the type, but just let the json parse it.

**Comment by taooceros** [comment]
It's a one time request when init, so probably handle it with a static template file is also fine.

**Comment by taooceros** [comment]
![image](https://user-images.githubusercontent.com/45326534/129605622-6c464b97-6411-4a3f-960d-b88e37645121.png)

**Comment by Garulf** [comment]
How does Flow handle secrets? Are they stored in the system keychain?

**Reply by jjw24** [reply]
there is no mechanism in Flow at the moment to encrypt/decrypt or store in system keychain, at the moment it's up to the plugin developer how to handle it, might be worth while adding in. Spotify and Github plugins both handles API token, last time i checked it's just stored as plain text for GitHub plugin locally, not sure about Spotify.

**Comment by taooceros** [comment]
This is on the way.

**Reply by Garulf** [reply]
So far looks very easy to implement and will be give a HUGE benefit to users. Fussing with text files is not a great user experience and will hopefully be a thing of the past with this feature.
